### ISSUE TITLE
problem with aggregated_by and mdtol and masks with iris2.1 and numpy1.15

### ISSUE THREAD

**Issue by mcguirepatr** [SciTools/iris#3190]
When I run these commands in python2.7 with iris 2.1 and numpy 1.15, I get the error messages in traceback below. 

PYTHON commands:
```python
import iris
cube = iris.load_cube('example.nc')
#check to see if the data is there:
cube.data
cube = cube.aggregated_by(["year", "day_of_year"], iris.analysis.MEAN, mdtol=0.03)
```

The 20MB example.nc NETCDF file is available here: 
https://drive.google.com/open?id=1hZNWXs7wOtBq2dXWM1zbAJnVMjy_DXc7

```python-traceback
Traceback (most recent call last):
  File "test10.py", line 10, in <module>
    cube = cube.aggregated_by(["year", "day_of_year"], iris.analysis.MEAN, mdtol=0.03)
  File "/Users/pmcguire/fluxnet1/iris/cube.py", line 3414, in aggregated_by
    **kwargs)
  File "/Users/pmcguire/fluxnet1/iris/analysis/__init__.py", line 524, in aggregate
    result.mask = result.mask | mask_update
  File "/Users/pmcguire/.local/lib/python2.7/site-packages/numpy/ma/core.py", line 6339, in __setattr__
    "attributes of {!r} are not writeable".format(self))
AttributeError: attributes of masked are not writeable
```
-------------------
I have tracked this down I think.

I could get test_fluxnet_evaluation.py to work if I comment out the three lines below from the _Aggregator(object).aggregate method in iris/analysis/__init__.py (near line 514) of a copy of iris2.1. 
The result.mask below was not defined for some reason even though ma.isMaskedArray(result) is true.
```python
      result = self.call_func(data, axis=axis, **kwargs)
        if (mdtol is not None and ma.isMaskedArray(data)):
            fraction_not_missing = data.count(axis=axis) / data.shape[axis]
            mask_update = 1 - mdtol > fraction_not_missing
            #if ma.isMaskedArray(result):
            #    result.mask = result.mask | mask_update
            #else:
            result = ma.array(result, mask=mask_update)
```
In version Iris1.13, this problem does not exist.
I can reproduce this problem if I upgrade from Iris1.13 to Iris2.2.0dev0, including all dependencies.

But then when I downgrade only Iris2.2 -> Iris.1.13, keeping the updated dependencies, the bug persists.
But then when I downgrade only Iris2.2 -> Iris.1.13 and numpy1.15->numpy1.12.1, keeping the updated dependencies, the bug goes away.
So I think it's something in numpy.

When it works with Iris1.13 and numpy1.12.1, Aggregrate (with iris.analysis.Mean) function tries to override the (non-array) mask value for the day, and it can do this, because the mask is writeable. But when it doesn't work with iris2.1 and numpy1.15, this function fails because this mask is not writeable.

2)A quick perusal of the numpy1.15 documentation suggests that they changed the mask behavior: https://docs.scipy.org/doc/numpy/release.html
"np.ma.masked is no longer writeable:
Attempts to mutate the masked constant now error, as the underlying arrays are marked readonly. In the past, it was possible to get away with...".
I am not sure if this is the change which is affecting us or not.

Can someone fix this problem in iris? Should I just maintain a separate iris branch (with those three lines commented out) for my own purposes till this is fixed? Or does anybody have a better solution? I am not sure if I should downgrade numpy or not.

Thanks,
Patrick McGuire

Department of Meteorology and Department of Geography & Environmental Science
University of Reading
Personal Web Page: http://www.personal.reading.ac.uk/~gn916173

**Comment by dkillick** [comment]
Hi @mcguirepatr - thanks for raising this issue report! As it happens I was just looking at your post on Google Groups; apologies that no-one has responded to your post there as yet.

I've successfully reproduced the issue you report here, and I think you're spot on in looking at NumPy as the cause of this issue, so we'll try and dig in and see if we can get a fix in place. In fact, your note about removing the mask update lines is very helpful for pointing towards a fix for this as it points to what the origins of the problem likely are. There are two:

* The NetCDF representation of data has recently changed such that all files are loaded with masked data by default, even if no masked points are present. This can cause a functionality problem in some cases, where the mask can be `False` (just a single value). I think this is happening here - but I need to check that.
* The line that reads `if ma.isMaskedArray(result):` should read `if ma.is_masked(result):`, as that provides a more general check that can deal with the presence of a mask, even if no masked points are set (that is, it can deal with situations like the one I outlined in the previous bullet point).

By the way, you don't need to call `cube.data`, as per the third line in your sample code. What this will do is load all data points off of disk and into memory. This has a performance impact and memory usage impact that increases with increased data sizes. Most operations in Iris are fully-functional (and automatically parallelised) without needing to load data in this way. Those that aren't will automatically load the data when it's needed.
(In connection to this point, I tried running your sample code without loading the data, and I noticed that Iris did not use the lazy aggregator in this case, which should probably also be investigated.)

P.S. I edited your initial post as there was some odd styling going on.

**Comment by dkillick** [comment]
> Iris did not use the lazy aggregator in this case

This is explained by #3129.

**Comment by mcguirepatr** [comment]
Thanks for your help with this, Mr. Killick! 
Did you figure out how to fix it?
Patrick McGuire

**Comment by darkblue-b** [comment]
as supporting info, I confirm that the example given appears CORRECTLY in OSGeoLive 12.0

`scitools-iris==2.1.0; Cartopy==0.14.2; cf-units==2.0.2
matplotlib==2.1.1; nc-time-axis==1.1.0
jupyter-core==4.4.0; jupyter-client==5.2.2
netCDF4==1.3.1; Cython==0.26.1; future==0.15.2; pyke==1.1.1; six==1.10.0
`

Ubuntu 1804;  Python 2.7;     iris_3190_bug_example.nc    21MB

**Comment by kwilliams-mo** [comment]
I'm also having this problem on the same system as Patrick - I think it's happening when one or more of the elements are masked after applying the aggregating function but before using mdtol.

Is there a work-around?

Thanks!

Example 1:

```
import numpy as np
import iris
cube = iris.cube.Cube(np.ma.MaskedArray([3]))

cube.add_dim_coord(iris.coords.DimCoord([1],long_name='foo'), (0,))

cube.aggregated_by('foo', iris.analysis.MEAN, mdtol=0.5) # no error

cube.data[0] = np.ma.masked

cube.aggregated_by('foo', iris.analysis.MEAN) # no error
cube.aggregated_by('foo', iris.analysis.MEAN, mdtol=0.5) # ValueError: assignment destination is read-only
```

Example 2:
```
import numpy as np
import iris

cube = iris.cube.Cube(np.arange(4) * 10.0)
cube.data = np.ma.asanyarray(cube.data)

dim_coord = iris.coords.DimCoord(np.arange(4),long_name='sample')
aux_coord = iris.coords.AuxCoord([2000, 2000, 2001, 2001], long_name='year')
cube.add_dim_coord(dim_coord, (0,))
cube.add_aux_coord(aux_coord, (0,))

cube.aggregated_by('year', iris.analysis.MEAN, mdtol=0.03) # no error

cube.data[0] = np.ma.masked
cube.aggregated_by('year', iris.analysis.MEAN, mdtol=0.03) # no error

cube.data[1] = np.ma.masked
cube.aggregated_by('year', iris.analysis.MEAN, mdtol=0.03) # ValueError: assignment destination is read-only
```

**Comment by kwilliams-mo** [comment]
This is what I've come up with so far as a work-around for Patrick's original example but it's pretty convoluted - there must be a better way! :)

```
import numpy as np
import iris
cube = iris.load_cube('~/example.nc')

try:
    agg_cube = cube.aggregated_by(["year", "day_of_year"], iris.analysis.MEAN, mdtol=0.03)

except ValueError:

    def make_doy_yyyy_arr(in_cube):
        '''
        Makes a list of the form ['DOY-YYYY', 'DOY-YYYY',....] using the 'day_of_year'
       and 'year' coordinates in a cube.
        '''
        mylist = [str(in_cube.coord('day_of_year').points[i]) + '-' + str(in_cube.coord('year').points[i]) for i in range(len(in_cube.coord('year').points))]
        return mylist

    # do the aggregation without mdtol
    agg_cube_no_mdtol_used = cube.aggregated_by(["year", "day_of_year"], iris.analysis.MEAN)

    # get rid of masked days (i.e. days with no valid data)
    agg_cube_no_mdtol_used = agg_cube_no_mdtol_used[~agg_cube_no_mdtol_used.data.mask]

    # get list of days which have at least one data point
    doy_yyyy_after_agg =  list(set(make_doy_yyyy_arr(agg_cube_no_mdtol_used)))

    # get list of 'DOY-YYYY' for each time step in original cube
    doy_yyyy_before_agg = make_doy_yyyy_arr(cube)

    # has_data indicates, for each timestep in the original cube, whether it's part of a day with at least some data
    has_data = []
    for i in range(len(doy_yyyy_before_agg)):
        if doy_yyyy_before_agg[i] in doy_yyyy_after_agg:
            has_data.append(True)
        else:
            has_data.append(False)

    # take out timesteps from days with no data
    new_cube = cube[has_data]

    # try the agggregation with mdtol again
    agg_cube = new_cube.aggregated_by(["year", "day_of_year"], iris.analysis.MEAN, mdtol=0.03)

print agg_cube
```

**Comment by rcomer** [comment]
I've reproduced the errors from @mcguirepatr's and @kwilliams-mo's examples with latest master branch (almost v3).  However, I note that the problem goes away if the cube's input data is lazy (lazy aggregated_by was introduced at Iris v2.4).  So one way to fix this would be to implement #2418.

**Comment by rcomer** [comment]
This is now fixed ready for Iris v3.3 (due out in the summer).  Sorry it took so long and thanks for your efforts in digging into the problem @mcguirepatr and @kwilliams-mo!
